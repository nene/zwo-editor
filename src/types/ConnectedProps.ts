type AnyFn = (...args: any) => any;

type ReturnVoid<F extends AnyFn> = (...args: Parameters<F>) => void;

type FunctionMap = { [key: string]: AnyFn };

type VoidFunctionMap<T extends FunctionMap> = {
  [P in keyof T]: ReturnVoid<T[P]>;
};

// Derives Props type from mapStateToProps and mapDispatchToProps
export type ConnectedProps<
  MapStateToProps extends AnyFn,
  MapDispatchToProps extends FunctionMap = {}
> = ReturnType<MapStateToProps> & VoidFunctionMap<MapDispatchToProps>;
