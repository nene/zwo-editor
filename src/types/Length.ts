export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;

// Alternative approach
//
// This way we could avoid storing class instances to Redux,
// which eliminates problems with serialization & deserialization.
/*
export type Distance = { meters: number };
export type Duration = { seconds: number };
export type Length = Distance | Duration;

export const Distance = (meters: number) => ({ meters });
export const Duration = (seconds: number) => ({ seconds });

export const isDistance = (obj: Length): obj is Distance => 'meters' in obj;
export const isDuration = (obj: Length): obj is Duration => 'seconds' in obj;
*/
