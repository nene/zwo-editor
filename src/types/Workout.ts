import { SportType } from "./SportType";
import { Interval } from "./Interval";
import { LengthType } from "./LengthType";

export interface Workout {
  author: string;
  name: string;
  description: string;
  sportType: SportType;
  lengthType: LengthType;
  tags: string[];
  intervals: Array<Interval>;
}

export function createEmptyWorkout(
  sportType: SportType,
  lengthType: LengthType
): Workout {
  return {
    author: "",
    name: "",
    description: "",
    sportType: sportType,
    lengthType: lengthType,
    tags: [],
    intervals: [],
  };
}
