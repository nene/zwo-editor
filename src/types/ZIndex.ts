export enum ZIndex {
  fader = 1,
  selected = 5,
  instruction = 11,
  tooltip = 100,
  popup = 1000,
}
