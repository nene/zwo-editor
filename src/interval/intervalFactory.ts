import { v4 as uuidv4 } from "uuid";
import { PaceType } from "../types/PaceType";
import {
  FreeInterval,
  Interval,
  RampInterval,
  RepetitionInterval,
  SteadyInterval,
} from "../types/Interval";
import { Distance, Duration, Length } from "../types/Length";
import { WorkoutMode } from "../modes/WorkoutMode";

const defaultDuration = Duration(300);
const defaultDistance = Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: "steady",
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: "ramp",
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: "free",
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: undefined,
      instructions: [],
      ...interval,
    };
  },

  repetition(
    interval: Partial<RepetitionInterval>,
    mode: WorkoutMode
  ): RepetitionInterval {
    return {
      type: "repetition",
      id: uuidv4(),
      repeat: 3,
      onLength: mode.lengthType === "time" ? Duration(30) : Distance(200),
      offLength: mode.lengthType === "time" ? Duration(120) : Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      onCadence: undefined,
      offCadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
