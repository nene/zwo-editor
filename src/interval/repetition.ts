import { range } from "ramda";
import { WorkoutMode } from "../modes/WorkoutMode";
import { RepetitionInterval, SteadyInterval } from "../types/Interval";
import { Length } from "../types/Length";
import intervalFactory from "./intervalFactory";

export function repetitions(
  repeat: number,
  onLength: Length,
  offLength: Length,
  interval: RepetitionInterval,
  mode: WorkoutMode
): SteadyInterval[] {
  return range(0, repeat).flatMap(() => [
    intervalFactory.steady(
      {
        length: onLength,
        intensity: interval.onIntensity,
        cadence: interval.onCadence,
        pace: interval.pace,
      },
      mode
    ),
    intervalFactory.steady(
      {
        length: offLength,
        intensity: interval.offIntensity,
        cadence: interval.offCadence,
        pace: interval.pace,
      },
      mode
    ),
  ]);
}
