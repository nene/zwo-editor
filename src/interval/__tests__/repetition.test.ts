import { repetitionToSteadyIntervals } from "../repetition";
import { Duration } from "../../types/Length";
import intervalFactory from "../intervalFactory";
import BikeMode from "../../modes/BikeMode";

jest.mock("uuid", () => ({
  v4: () => `mock-id`,
}));

const defaultBikeMode = () => new BikeMode(200, 75);

describe("repetitionToSteadyIntervals()", () => {
  it("converts RepetitionInterval to an array of SteadyIntervals", () => {
    const mode = defaultBikeMode();
    expect(
      repetitionToSteadyIntervals(
        intervalFactory.repetition(
          {
            id: "#1",
            repeat: 2,
            onLength: Duration(60),
            offLength: Duration(30),
            onIntensity: 0.9,
            offIntensity: 0.75,
            onCadence: 90,
            offCadence: undefined,
          },
          mode
        ),
        mode
      )
    ).toMatchSnapshot();
  });
});
