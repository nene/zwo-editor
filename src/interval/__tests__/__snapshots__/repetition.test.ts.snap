// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repetitionToSteadyIntervals() converts RepetitionInterval to an array of SteadyIntervals 1`] = `
Array [
  Object {
    "cadence": 90,
    "id": "mock-id",
    "instructions": Array [],
    "intensity": 0.9,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "mock-id",
    "instructions": Array [],
    "intensity": 0.75,
    "length": Object {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 90,
    "id": "mock-id",
    "instructions": Array [],
    "intensity": 0.9,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "mock-id",
    "instructions": Array [],
    "intensity": 0.75,
    "length": Object {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`repetitionToSteadyIntervals() distributes instructions between SteadyIntervals 1`] = `
Array [
  Object {
    "cadence": 90,
    "id": "mock-id",
    "instructions": Array [
      Object {
        "id": "mock-id",
        "offset": Object {
          "seconds": 10,
        },
        "text": "In first interval",
      },
    ],
    "intensity": 0.9,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "mock-id",
    "instructions": Array [
      Object {
        "id": "mock-id",
        "offset": Object {
          "seconds": 60,
        },
        "text": "In second interval #1",
      },
      Object {
        "id": "mock-id",
        "offset": Object {
          "seconds": 80,
        },
        "text": "In second interval #2",
      },
    ],
    "intensity": 0.75,
    "length": Object {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 90,
    "id": "mock-id",
    "instructions": Array [
      Object {
        "id": "mock-id",
        "offset": Object {
          "seconds": 100,
        },
        "text": "In third interval",
      },
    ],
    "intensity": 0.9,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "mock-id",
    "instructions": Array [
      Object {
        "id": "mock-id",
        "offset": Object {
          "seconds": 150,
        },
        "text": "In fourth interval",
      },
    ],
    "intensity": 0.75,
    "length": Object {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;
